hadoop-mapreduce-project/hadoop-mapreduce-client/hadoop-mapreduce-client-core/src/main/java/org/apache/hadoop/mapreduce/lib/output/committer/manifest/stages/CreateOutputDirectoriesStage.java

```diff
@@ -0,0 +1,470 @@
+        getName(), count, path, e.toString());
+    LOG.debug("{}: Full exception details",
+        getName(), e);
+    if (count == 1  && !getStageConfig().getPrepareParentDirectories()) {
+      // first failure and no parent dir setup, so recommend
+      // preparing the parent dir
+      LOG.warn(HINT_PREPARE_PARENT_DIR);
+    }
+  }
+
+  /**
+   * Clean up parent dirs. No attempt is made to create them, because
+   * we know mkdirs() will do that (assuming it has the permissions,
+   * of course)
+   * @param directoriesToCreate set of dirs to create
+   * @throws IOException IO problem
+   */
+  private void prepareParentDirectories(final Set<Path> directoriesToCreate)
+      throws IOException {
+
+    // include parents in the paths
+    final Set<Path> ancestors = extractAncestors(
+        getStageConfig().getDestinationDir(),
+        directoriesToCreate);
+
+    final int ancestorCount = ancestors.size();
+    LOG.info("Preparing {} ancestor directory/directories", ancestorCount);
+    // if there is a single ancestor, don't bother with parallisation.
+    Duration d = measureDurationOfInvocation(getIOStatistics(),
+        OP_PREPARE_DIR_ANCESTORS, () ->
+            TaskPool.foreach(ancestors)
+                .executeWith(getIOProcessors(ancestorCount))
+                .stopOnFailure()
+                .run(dir -> {
+                  updateAuditContext(OP_PREPARE_DIR_ANCESTORS);
+                  prepareParentDir(dir);
+                }));
+    LOG.info("Time to prepare ancestor directories {}", humanTime(d.toMillis()));
+  }
+
+  /**
+   * Prepare a parent directory.
+   * @param dir directory to probe
+   * @throws IOException failure in probe other than FNFE
+   */
+  private void prepareParentDir(Path dir) throws IOException {
+    // report progress back
+    progress();
+    LOG.debug("{}: Probing parent dir {}", getName(), dir);
+    if (isFile(dir)) {
+      // it's a file: delete it.
+      LOG.info("{}: Deleting file {}", getName(), dir);
+      delete(dir, false, OP_DELETE_FILE_UNDER_DESTINATION);
+      // note its final state
+      addToDirectoryMap(dir, DirMapState.ancestorWasFileNowDeleted);
+    } else {
+      // and add to dir map as a dir or missing entry
+      LOG.debug("{}: Dir {} is missing or a directory", getName(), dir);
+      addToDirectoryMap(dir, DirMapState.ancestorWasDirOrMissing);
+    }
+  }
+
+  /**
+   * Determine all ancestors of the dirs to create which are
+   * under the dest path, not in the set of dirs to create.
+   * Static and public for testability.
+   * @param destDir destination dir of the job
+   * @param directoriesToCreate set of dirs to create from tasks.
+   * @return the set of parent dirs
+   */
+  public static Set<Path> extractAncestors(
+      final Path destDir,
+      final Collection<Path> directoriesToCreate) {
+
+    final Set<Path> ancestors = new HashSet<>(directoriesToCreate.size());
```

Comment:
In what scenario would the directories to create contain a path outside of the destination directory?
If I'm not mistaken, the destination directory is the configured output directory ... and the source directory will always be something under $outputDir/.temporary/*
Wondering if for safety - this should error-out if the target is outside of the destination directory

Reply:
More specifically - if we ever end up trying to delete a directory in this list.